packages/blocks/src/api/parser.js

```diff
@@ -262,9 +292,9 @@ export function getBlockAttribute( attributeKey, attributeSchema, innerHTML, com
 			break;
 	}
 
-	if ( type !== undefined && ! isOfTypes( value, castArray( type ) ) ) {
-		// Reject the value if it is not valid of type. Reverting to the
-		// undefined value ensures the default is restored, if applicable.
+	if ( ! isValidByType( value, type ) || ! isValidByEnum( value, enumSet ) ) {
+		// Reject the value if it is not valid. Reverting to the undefined
```

Comment:
Should it warn in development mode? I guess the answer should be yes. What's the reason we silently reject it? I don't see it as a blocker but I would like to see a roadmap for making extensibility easier by providing good feedback for developers.

Reply:
> Should it warn in development mode? I guess the answer should be yes. What's the reason we silently reject it? I don't see it as a blocker but I would like to see a roadmap for making extensibility easier by providing good feedback for developers.

I agree on all accounts. I'm inclined to merge this as-is with consideration that it follows the same precedent established by type validation. I agree we should probably include some feedback here, and will plan to create a follow-up issue.

Reply:
In fact, I think it's quite similar to the issue at #7653 , which had been closed (as it impacted the _server_ implementation, which we're mirroring now in the client).

Which is worth highlighting: Server-side we silently drop the attributes as well:

https://github.com/WordPress/WordPress/blob/2887cea50f2a717782a6adb5922cb272d3c4bc10/wp-includes/class-wp-block-type.php#L153-L156

Whatever we choose to do, I'd suggest it be made consistent across the two runtimes.

Do you think it'd make sense to reopen #7653 for this? Or create a separate issue?

Reply:
We rather open a ticket in Trac and new issue on GitHub as those are related things but still different. I think that validation and error handling in Gutenberg are rather randomly added and it should be something that should become the focus. Ideally, there is RFC created with a project-wide proposal rather than trying to fix it case by case.

There is this related work needed to implement errors catching in code registered by plugins:
* `withSelect`
* `withDispatch`
* `withFilters`
* `PluginArea`

Reply:
Hi @gziolo , @aduth ,

I'm new to WordPress and was debugging someone else's custom block that was failing validation on reloading the editor. The silent failure here made it very difficult to understand why the block would fail validation upon reloading the editor and I ended up having to use a debugger and trace/read the Gutenberg source back to this section of the code. Was a ticket ever created to track better logging/feedback for errors? I haven't been able to find it.

---

Longer explanation on why I feel this is needed:

I had initially missed [this bit of explanation in the Block documentation](https://developer.wordpress.org/block-editor/developers/block-api/block-attributes/#considerations):

> Lastly, make sure that you respect the data’s type when setting attributes, as the framework does not automatically perform type casting of meta. Incorrect typing in block attributes will result in a post remaining dirty even after saving

Combined with the silent validation failures made me believe that the information was indeed being saved as expected. I validated this by confirming that the expected data was saved in `wp_posts`. Since I was in development mode and saw no other errors/warnings, there was no good place for a dev to being investigation. Any amount of feedback while developing would have drastically cut my investigation time.

Also, with regards to the Block documentation, I would argue that if no validation feedback will be added, then the above explanation should be moved out of **Considerations** and more prominently explained at the top of the **Attributes** section as it's a requirement to making custom blocks work.